### PR TITLE
Add error match pattern 'var reason' validation

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
$subject.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/18989 and https://github.com/ballerina-platform/ballerina-lang/issues/18997

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 1]
```diff
@@ -634,6 +634,9 @@ error.undefined.parameter=\
 error.invalid.error.reason.type=\
   invalid error reason type ''{0}'', expected a subtype of ''string''
 
+error.error.match.over.const.reason.ref.not.supported=\
+  error match pattern with reason as a ''constant'' reference is currently not supported
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]:
```suggestion
  error match pattern with a constant reference as the reason is not yet supported
```

Comment (pubudu91) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: How about something like 
```
matching error reason to a ''const'' pattern is currently not supported.
```
?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/structured_error_match_patterns.bal [hunk 2]
```diff
@@ -329,16 +329,28 @@ function testErrorConstReasonMatchPattern() returns string {
...
             return reason;
         }
     }
     return "Default";
 }
+
+public function testErrorMatchWihtoutReason() returns string {
+    string|error se = error("test reason", message="error detail message");
+
+    match se {
```
Comment (pubudu91) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/structured_error_match_patterns.bal]: Does the wild card match pattern work for the error reason? If so, shall we add a test case for that as well?

File: compiler/ballerina-lang/src/main/resources/compiler.properties (hunk 1 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/resources/compiler.properties]: Matching to a const literal is supported

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/structured_error_match_patterns.bal (hunk 2 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/structured_error_match_patterns.bal]: I think it's already there https://github.com/ballerina-platform/ballerina-lang/blob/master/tests/jballerina-unit-test/src/test/resources/test-src/statements/matchstmt/structured_error_match_patterns.bal#L151

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/matchstmt/MatchStructuredErrorPatternsTest.java [hunk 3]
```diff
@@ -199,4 +199,22 @@ public void testUnreachablePatterns() {
...
         BAssertUtil.validateError(resultNegative, ++i, unreachablePattern, 85, 13);
     }
+
+    @Test(description = "Test unsupported error match pattern")
+    public void testErrorMatchPatternNotSupportedErrors() {
+        CompileResult result = BCompileUtil.compile(
+                "test-src/statements/matchstmt/structured_error_match_patterns_negative2.bal");
+        int i = 0;
+        BAssertUtil.validateError(result, i++,
+                "error match pattern with reason as a 'constant' reference is currently not supported", 33, 15);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/matchstmt/MatchStructuredErrorPatternsTest.java]: tests needs to be updated for the error message change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 4]
```diff
@@ -1353,6 +1353,22 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable) {
                 errorVariable.restDetail.type = restType;
                 errorVariable.restDetail.accept(this);
             }
+
+            if (isReasonSpecified(errorVariable)
+                    && !errorVariable.reasonVarPrefixAvailable
+                    && errorVariable.reasonMatchConst == null
+                    && errorVariable.isInMatchStmt) {
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Do we need a new field btw? Haven't we come across the requirement to identify if we are in a match statement before?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/matchstmt/MatchStructuredErrorPatternsTest.java (hunk 3 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/matchstmt/MatchStructuredErrorPatternsTest.java]: Yes, I forgot that

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: > Do we need a new field btw? Haven't we come across the requirement to identify if we are in a match statement before?

AFAIK we don't support const matching in other match patterns and treat all names as a variable to be bound to members of matched value. I think that's why didn't wanted something like this before.

[the remark this answers is not in the document]
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: > Shouldn't this be the first check?

Will move to top as it makes more sense to check reason, detail, rest in that order.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 5]
```diff
@@ -1318,6 +1318,22 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable) {
             Map<String, BField> fieldMap = recordType.fields.stream()
                     .collect(Collectors.toMap(f -> f.name.value, f -> f));
 
+            if (isReasonSpecified(errorVariable)
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: We can move this to L1317 right? The variables defined are not used in this block?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 6]
```diff
@@ -1318,6 +1318,22 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable) {
...
+                    && !errorVariable.reasonVarPrefixAvailable
+                    && errorVariable.reasonMatchConst == null
+                    && errorVariable.isInMatchStmt) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: What I originally meant is whether `errorVariable.isInMatchStmt` should be the first check in the if block?

Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: No preferences from me, will move up

**Pull request merged**